### PR TITLE
[OpenCL] Make extension function lookup return unusupported error

### PR DESCRIPTION
**Pull request by steffenlarsen**
This commit changes the error code of failing to look up extension pointers in OpenCL return UR_RESULT_ERROR_UNSUPPORTED_FEATURE instead of UR_RESULT_ERROR_INVALID_VALUE. Likewise, some of the places this lookup was used, error would return UR_RESULT_ERROR_INVALID_OPERATION. The motivation behind this change is to make the error codes for cases where extensions are not supported for the OpenCL target clearer.

### CODE REVIEW

File: source/adapters/opencl/command_buffer.cpp [hunk 1]
```diff
@@ -26,8 +26,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferCreateExp(
           CLContext, cl_ext::ExtFuncPtrCache->clCreateCommandBufferKHRCache,
           cl_ext::CreateCommandBufferName, &clCreateCommandBufferKHR);
 
-  if (!clCreateCommandBufferKHR || Res != CL_SUCCESS)
-    return UR_RESULT_ERROR_INVALID_OPERATION;
+  if (Res != CL_SUCCESS)
+    return Res;
```
Comment (aarongreig) [inline source/adapters/opencl/command_buffer.cpp:30]: this is returning a CL error code where there should be a UR one, instead we should remove this altogether and wrap the `getExtFunc` calls in `UR_RETURN_ON_FAILURE` (applies elsewhere)

**Comment by aarongreig** [comment]
relates to #1362

File: source/adapters/opencl/command_buffer.cpp (hunk 1 above)
Comment (steffenlarsen) [reply source/adapters/opencl/command_buffer.cpp]: Good eye! This should have been addressed.

**Review by EwanC — APPROVED** [review]

**Comment by steffenlarsen** [comment]
> relates to #1362

Interesting! This was an ambiguity we ran into when debugging an issue with an unexpected device being picked up in DPC++. I have made a patch to also make a mapping for the unsupported error in intel/llvm as well: https://github.com/intel/llvm/pull/13036

The hope is that this will make it easier to differentiate when the backend is unhappy with a call and when it is simply just not supporting it.

File: source/adapters/opencl/enqueue.cpp [hunk 2]
```diff
@@ -409,21 +403,21 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueReadHostPipe(
...
-  ur_result_t RetVal =
+  UR_RETURN_ON_FAILURE(
       cl_ext::getExtFuncFromContext<cl_ext::clEnqueueReadHostPipeINTEL_fn>(
           CLContext, cl_ext::ExtFuncPtrCache->clEnqueueReadHostPipeINTELCache,
-          cl_ext::EnqueueReadHostPipeName, &FuncPtr);
+          cl_ext::EnqueueReadHostPipeName, &FuncPtr));
 
   if (FuncPtr) {
-    RetVal = mapCLErrorToUR(
+    CL_RETURN_ON_FAILURE(mapCLErrorToUR(
```
Comment (aarongreig) [inline source/adapters/opencl/enqueue.cpp:412]:
```suggestion
    CL_RETURN_ON_FAILURE(
```
`CL_RETURN_ON_FAILURE` does `mapCLErrorToUR` for you (same [here](https://github.com/oneapi-src/unified-runtime/pull/1448/files#diff-c80bbd3e96dbcb2b2ca1d99af816a310b1720decee46fdb0e97d5fd3075e5cb3R444))

**Review by aarongreig — APPROVED** [review]

**Comment by steffenlarsen** [comment]
@kbenzie - I think this is ready.

**Comment by hvdijk** [comment]
Hi, there were some checks for functions that were followed by a `if (FuncPtr)` check and properly handled the optional function not being implemented. It looks like they now return an error for that, is that right? If I am reading this right (and I may well not, please correct me if so) this change basically means that OpenCL implementations other than Intel's own can no longer be used. I'm seeing this with OCK where we now get e.g. when running SYCL-CTS
```
---> piMemBufferCreate(
	<unknown> : 0xa204c80
	<unknown> : 9
	<unknown> : 40
	<unknown> : 0x7ffff86c0910
	<unknown> : 0x7ffff86bf5a0
	<unknown> : 0
) ---> 	pi_result : -995
	[out]void * : 0x7ffff86c0910
	[out]pi_mem * : 0x7ffff86bf5a0[ 0 ... ]


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The SYCL 2020 Conformance Test Suite is a Catch2 v3.2.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
requisite for the entire underlying buffer for sycl::accessor 
-------------------------------------------------------------------------------
/builds/ComputeAorta/ci-github/SYCL-CTS/tests/accessor/accessor_requisite_entire_buffer.cpp:32
...............................................................................

/builds/ComputeAorta/ci-github/SYCL-CTS/tests/accessor/accessor_requisite_entire_buffer.cpp:32: FAILED:
due to unexpected exception with message:
  SYCL exception
  with category name: 'sycl'
  with code: 'runtime'
  with code raw value: '1'
  with code message: 'SYCL Error'
  with what: 'Native API failed. Native API returns: -995 (The plugin or device
  does not support the called function) -995 (The plugin or device does not
  support the called function)'

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```

**Comment by coldav** [comment]
note that OCK was passing (fast) SYCL CTS before this point, so it suggests that perhaps we have made something optional an error.

**Comment by steffenlarsen** [comment]
Thank you for letting me know, @hvdijk & @coldav ! You are absolutely right, the changes to {{urMemBufferCreate}} were not right. They can be safely reverted, as done in https://github.com/oneapi-src/unified-runtime/pull/1496.

**Pull request merged**